app/src/main/java/com/example/chatbotandroid/controller/MainActivity.kt

```diff
@@ -0,0 +1,108 @@
+import android.os.Bundle
+import com.example.chatbotandroid.R
+import androidx.recyclerview.widget.LinearLayoutManager
+import com.example.chatbotandroid.model.ChatAdapter
+import com.example.chatbotandroid.model.Message
+import com.example.chatbotandroid.utils.Constants
+import com.example.chatbotandroid.utils.Time
+import org.java_websocket.client.WebSocketClient
+import kotlinx.android.synthetic.main.activity_main.*
+import kotlinx.coroutines.*
+import org.java_websocket.handshake.ServerHandshake
+import java.lang.Exception
+import java.net.URI
+
+class MainActivity : AppCompatActivity() {
+    private lateinit var adapter: ChatAdapter
+    private lateinit var webSocketClient: WebSocketClient
+    private val webSocketUri: URI? = URI(Constants.WEB_SOCKET_URL)
+
+    //Life-cycle methods
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        recyclerView()
+        registerOnClickListeners()
+    }
+
+    override fun onResume() {
+        super.onResume()
+        initWebSocket()
+    }
+
+    override fun onPause() {
+        super.onPause()
+        //Close web socket if app is on background to avoid unnecessary active connection
+        webSocketClient.close()
+    }
+
+    //Websocket methods
+    private fun initWebSocket() {
+        createWebSocketClient(webSocketUri)
+        webSocketClient.connect()
+    }
+
+    private fun createWebSocketClient(wsURI: URI?) {
+        webSocketClient = object : WebSocketClient(wsURI) {
+            override fun onOpen(handshakedata: ServerHandshake?) {
+                println("Websocket connection open")
+            }
+            override fun onMessage(message: String?) {
+                if (message != null) {
+                    appendServerResponse(message)
+                }
+            }
+            override fun onClose(code: Int, reason: String?, remote: Boolean) {
+                println(reason)
+            }
+            override fun onError(ex: Exception?) {
+                println(ex)
+            }
+        }
+    }
+
+    //Helper methods
+    private fun registerOnClickListeners(){
+        //Send message from client to server
+        send_button.setOnClickListener {
+            sendMessage()
+        }
+    }
+
+    private fun sendMessage() {
+        val message = input_message.text.toString()
+
+        if (message.isNotEmpty()) {
+            //Clean message input
+            input_message.setText("")
+            //Pass message to adapter to add it to chat view
+            adapter.insertMessage(Message(message, Constants.SEND_TYPE, Time.timeStamp()))
+            recycler_view_chat.scrollToPosition(adapter.itemCount - 1)
+            webSocketClient.send(message)
+        }
+    }
+
+    private fun appendServerResponse(message: String) {
+        GlobalScope.launch {
+            //Simulate network latency
+            delay(500)
```

Comment:
great idea. this kind of asynchronous work would also do nicely to live inside a `ViewModel` (see my other comment about `ViewModel`s), because of the way the Activity lifecycle works. that is, if you launch a coroutine here, it's holding an outer reference to MainActivity (that's how it can access things like the `MainActivity.adapter` field). if the Activity were recreated within this delay period, e.g. due to a configuration change, then the coroutine would be holding onto and trying to access an Activity that was destroyed (but still in memory - leaked by this coroutine) once the delay is over. best case scenario is nothing would get updated, worst case is an exception could be thrown.